File: system/stackswap.go
// Copyright 2015 The dpar Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package system

import (
	"errors"
	"fmt"
	"strings"
)

// Archetype transitions + interface validation.
var archetypeSSWShift Transition = sswShift{}
var archetypeSSWSwap Transition = sswSwap{}
var archetypeSSWLeftArc Transition = sswLeftArc{"<archetype>"}
var archetypeSSWRightArc Transition = sswRightArc{"<archetype>"}

// Assert TransitionSystem/TransitionSerializer conformance.
var _ TransitionSystem = NewStackSwap()
var _ TransitionSerializer = NewStackSwap()

// The StackSwap transition-system is similar to the StackProjective system,
// but adds a swap transition that allows it to produce non-projective trees.
//
// This system is described in:
//
// Joakim Nivre, Non-projective dependency parsing in expected linear time, 2009
type StackSwap struct {
	archetypeTransitions TransitionSet
}

// NewStackSwap creates a stack-swap transition system.
func NewStackSwap() *StackSwap {
	trans := map[Transition]interface{}{
		archetypeSSWShift:    nil,
		archetypeSSWSwap:     nil,
		archetypeSSWLeftArc:  nil,
		archetypeSSWRightArc: nil,
	}

	return &StackSwap{trans}
}

// IsTerminal checks whether a parser configuration is a terminal
// configuration for the system. When this is the case, no further
// transitions should be attempted.
func (ts *StackSwap) IsTerminal(c Configuration) bool {
	return len(c.Stack) == 1 && len(c.Buffer) == 0
}

// PossibleTransitions returns the set of possible transitions in a
// particular parser configuration. The returned transitions are
// archetypal transitions (they do not parametrized over a dependency
// relation).
func (ts *StackSwap) PossibleTransitions(configuration Configuration) TransitionSet {
	possible := make(TransitionSet)

	for trans := range ts.archetypeTransitions {
		if trans.IsPossible(configuration) {
			possible[trans] = nil
		}
	}

	return possible
}

// SerializeTransition serializes a transition of the stack-swap
// system to a string. An error is returned when it is attempted to
// serialize a transition of another system.
func (ts *StackSwap) SerializeTransition(t Transition) (string, error) {
	switch t := t.(type) {
	case sswLeftArc:
		return fmt.Sprintf("LEFT_ARC %s", t.relation), nil
	case sswRightArc:
		return fmt.Sprintf("RIGHT_ARC %s", t.relation), nil
	case sswSwap:
		return "SWAP", nil
	case sswShift:
		return "SHIFT", nil
	default:
		return "", fmt.Errorf("Not a transition of the stack-swap system: %T", t)
	}
}

// DeserializeTransition deserializes a transition from the
// stack-swap system. If the provided transition is not known to
// the system, an error is returned.
func (ts *StackSwap) DeserializeTransition(transString string) (Transition, error) {
	parts := strings.SplitN(transString, " ", 2)

	if len(parts) == 0 {
		return nil, errors.New("Empty transition")
	}

	switch parts[0] {
	default:
		return nil, fmt.Errorf("Unknown transition: %s", parts[0])
	case "LEFT_ARC":
		if len(parts) == 1 {
			return nil, errors.New("Left-Arc transition requires label argument")
		}
		return sswLeftArc{parts[1]}, nil
	case "RIGHT_ARC":
		if len(parts) == 1 {
			return nil, errors.New("Right-Arc transition requires label argument")
		}
		return sswRightArc{parts[1]}, nil
	case "SWAP":
		return sswSwap{}, nil
	case "SHIFT":
		return sswShift{}, nil
	}
}

type sswLeftArc struct {
	relation string
}

func (l sswLeftArc) IsPossible(c Configuration) bool {
	stackSize := len(c.Stack)
	return stackSize > 1 && c.Stack[stackSize-2] != 0
}

func (l sswLeftArc) Apply(c *Configuration) {
	stack := c.Stack

	stackSize := len(stack)
	head := stack[stackSize-1]
	dependent := stack[stackSize-2]
	dependency := Dependency{head, l.relation, dependent}

	c.AddDependency(&dependency)

	c.Stack = append(stack[:stackSize-2], head)
}

type sswRightArc struct {
	relation string
}

func (r sswRightArc) IsPossible(configuration Configuration) bool {
	return len(configuration.Stack) > 1
}

func (r sswRightArc) Apply(c *Configuration) {
	stack := c.Stack

	stackSize := len(stack)
	dependent := stack[stackSize-1]
	head := stack[stackSize-2]
	dependency := Dependency{head, r.relation, dependent}

	c.AddDependency(&dependency)

	c.Stack = stack[:stackSize-1]
}

type sswSwap struct{}

func (s sswSwap) IsPossible(c Configuration) bool {
	stack := c.Stack
	stackSize := len(stack)
	return stackSize > 1 && stack[stackSize-1] > stack[stackSize-2]
}

func (s sswSwap) Apply(c *Configuration) {
	stack := c.Stack
	stackSize := len(stack)

	// Move second stack element to the buffer...
	// Todo: exploit possible gap at the beginning of the backing array...
	newBuffer := make([]uint, len(c.Buffer)+1)
	copy(newBuffer[1:], c.Buffer)
	newBuffer[0] = stack[stackSize-2]
	c.Buffer = newBuffer

	// ... and remove that element from the stack.
	head := stack[stackSize-1]
	c.Stack = append(stack[:stackSize-2], head)
}

type sswShift struct{}

func (s sswShift) IsPossible(c Configuration) bool {
	return len(c.Buffer) != 0
}

func (s sswShift) Apply(c *Configuration) {
	if !s.IsPossible(*c) {
		panic("Tried to shift on an empty buffer.")
	}

	stack := c.Stack
	buffer := c.Buffer
	c.Stack = append(stack, buffer[0])
	c.Buffer = buffer[1:]
}

// The StackSwapOracle returns a correct transition for a particular
// configuration in the stack-swap system. The transition choice in
// an oracle is based on the gold-standard parse.
type StackSwapOracle struct {
	dependentHeadMapping map[uint]Dependency
	projectiveOrder      []int
}

// NewStackSwapOracle returns a new oracle from the stack-swap
// system. The gold dependency set is used to choose a correct transition
// from each configuration.
func NewStackSwapOracle(goldDependencies DependencySet) Guide {
	oracle := StackSwapOracle{
		dependentHeadMapping: goldDependencies.CreateDependentHeadMapping(),
		projectiveOrder:      extractProjectiveOrder(goldDependencies),
	}
	return &oracle
}

func extractProjectiveOrder(dependencies DependencySet) []int {
	g := dependenciesToGraph(dependencies)

	projectiveOrder := make([]uint, 0, len(dependencies)+1)

	dfs(g, 0, func(token int) {
		projectiveOrder = append(projectiveOrder, uint(token))
	})

	orderMapping := make([]int, len(projectiveOrder))
	for idx, token := range projectiveOrder {
		orderMapping[token] = idx
	}

	return orderMapping
}

// BestTransition returns the best transition from the current
// configuration, as inferred from the gold standard dependency
// structure.
func (o *StackSwapOracle) BestTransition(c *Configuration) Transition {
	stack := c.Stack

	if len(stack) > 1 {
		stackSize := len(stack)
		stack0 := stack[stackSize-1]
		stack1 := stack[stackSize-2]

		la := sswLeftArc{o.dependentHeadMapping[stack1].Relation}
		if la.IsPossible(*c) && o.dependentHeadMapping[stack1].Head == stack0 && !o.neededForAttachment(c, stack1) {
			return la
		}

		ra := sswRightArc{o.dependentHeadMapping[stack0].Relation}
		if ra.IsPossible(*c) &&
			o.dependentHeadMapping[stack0].Head == stack1 &&
			!o.neededForAttachment(c, stack0) {
			return ra
		}

		if o.projectiveOrder[stack0] < o.projectiveOrder[stack1] {
			return sswSwap{}
		}
	}

	return sswShift{}
}

func (o *StackSwapOracle) neededForAttachment(c *Configuration, token uint) bool {
	for _, bufToken := range c.Buffer {
		if o.dependentHeadMapping[bufToken].Head == token {
			return true
		}
	}

	for _, bufToken := range c.Stack {
		if o.dependentHeadMapping[bufToken].Head == token {
			return true
		}
	}

	return false
}

File: dpar/src/systems/stack_swap.rs
use std::collections::HashMap;

use petgraph::graph::node_index;
use petgraph::visit::Dfs;
use petgraph::{Directed, Graph};

use guide::Guide;
use system::{
    Dependency, DependencySet, ParserState, Transition, TransitionLookup, TransitionSystem,
};

use features::addr::Source;
use system::AttachmentAddr;
use systems::util::dep_head_mapping;

/// The stack-swap transition system for non-projective parsing.
///
/// The stack-swap transition-system is similar to the stack-projective system,
/// but adds a swap transition that allows it to produce non-projective trees.
///
/// This system is described in:
///
/// Joakim Nivre, Non-projective dependency parsing in expected linear time, 2009
#[derive(Eq, PartialEq, Serialize, Deserialize)]
pub struct StackSwapSystem {
    transitions: TransitionLookup<StackSwapTransition>,
}

impl StackSwapSystem {
    pub fn new() -> Self {
        StackSwapSystem {
            transitions: TransitionLookup::default(),
        }
    }
}

impl Default for StackSwapSystem {
    fn default() -> Self {
        StackSwapSystem::new()
    }
}

impl TransitionSystem for StackSwapSystem {
    type Transition = StackSwapTransition;
    type Oracle = StackSwapOracle;

    const ATTACHMENT_ADDRS: [AttachmentAddr; 2] = [
        AttachmentAddr {
            head: Source::Stack(0),
            dependent: Source::Stack(1),
        },
        AttachmentAddr {
            head: Source::Stack(1),
            dependent: Source::Stack(0),
        },
    ];

    fn is_terminal(state: &ParserState) -> bool {
        state.buffer().is_empty() && state.stack().len() == 1
    }

    fn oracle(gold_dependencies: &DependencySet) -> Self::Oracle {
        StackSwapOracle::new(gold_dependencies)
    }

    fn transitions(&self) -> &TransitionLookup<Self::Transition> {
        &self.transitions
    }
}

/// Stack-projective transition.
#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub enum StackSwapTransition {
    LeftArc(String),
    RightArc(String),
    Shift,
    Swap,
}

impl Transition for StackSwapTransition {
    type S = StackSwapSystem;

    fn is_possible(&self, state: &ParserState) -> bool {
        let stack_len = state.stack().len();

        match self {
            &StackSwapTransition::LeftArc(_) => stack_len > 2,
            &StackSwapTransition::RightArc(_) => stack_len > 1,
            &StackSwapTransition::Shift => !state.buffer().is_empty(),
            &StackSwapTransition::Swap => {
                stack_len > 2 && state.stack()[stack_len - 1] > state.stack()[stack_len - 2]
            }
        }
    }

    fn apply(&self, state: &mut ParserState) {
        let stack_len = state.stack().len();

        match self {
            &StackSwapTransition::LeftArc(ref rel) => {
                let head = state.stack()[stack_len - 1];
                let dependent = state.stack_mut().remove(stack_len - 2);

                state.add_dependency(Dependency {
                    head: head,
                    relation: rel.clone(),
                    dependent: dependent,
                });
            }
            &StackSwapTransition::RightArc(ref rel) => {
                let head = state.stack()[stack_len - 2];
                let dependent = state.stack_mut().remove(stack_len - 1);

                state.add_dependency(Dependency {
                    head: head,
                    relation: rel.clone(),
                    dependent: dependent,
                });
            }
            &StackSwapTransition::Shift => {
                let next = state.buffer_mut().remove(0);
                state.stack_mut().push(next);
            }
            &StackSwapTransition::Swap => {
                let swap_token = state.stack_mut().remove(stack_len - 2);
                state.buffer_mut().insert(0, swap_token);
            }
        }
    }
}

pub struct StackSwapOracle {
    dependencies: HashMap<usize, Dependency>,
    projective_order: Vec<usize>,
}

impl StackSwapOracle {
    pub fn new(gold_dependencies: &DependencySet) -> StackSwapOracle {
        StackSwapOracle {
            dependencies: dep_head_mapping(gold_dependencies),
            projective_order: extract_projective_order(gold_dependencies),
        }
    }

    fn needed_for_attachment(&self, state: &ParserState, token: usize) -> bool {
        for buf_token in state.buffer() {
            if let Some(dep) = self.dependencies.get(&buf_token) {
                if dep.head == token {
                    return true;
                }
            }
        }

        for stack_token in state.stack() {
            if let Some(dep) = self.dependencies.get(&stack_token) {
                if dep.head == token {
                    return true;
                }
            }
        }

        false
    }
}

impl Guide for StackSwapOracle {
    type Transition = StackSwapTransition;

    fn best_transition(&mut self, state: &ParserState) -> StackSwapTransition {
        let stack = &state.stack();
        let stack_len = stack.len();

        if stack_len > 1 {
            let stack0 = stack[stack_len - 1];
            let stack1 = stack[stack_len - 2];

            if let Some(dep) = self.dependencies.get(&stack1) {
                let la = StackSwapTransition::LeftArc(dep.relation.clone());

                if dep.head == stack0
                    && la.is_possible(state)
                    && !self.needed_for_attachment(state, stack1)
                {
                    return la;
                }
            }

            if let Some(dep) = self.dependencies.get(&stack0) {
                let ra = StackSwapTransition::RightArc(dep.relation.clone());

                if dep.head == stack1
                    && ra.is_possible(state)
                    && !self.needed_for_attachment(state, stack0)
                {
                    return ra;
                }
            }

            if self.projective_order[stack0] < self.projective_order[stack1] {
                return StackSwapTransition::Swap;
            }
        }

        StackSwapTransition::Shift
    }
}

fn dependencies_to_graph(dependencies: &DependencySet) -> Graph<(), (), Directed> {
    let edges = dependencies.iter().map(|dependency| {
        (
            node_index(dependency.head),
            node_index(dependency.dependent),
        )
    });

    Graph::from_edges(edges)
}

fn extract_projective_order(dependencies: &DependencySet) -> Vec<usize> {
    // First get the tokens in projective order.
    let mut projective_order = Vec::with_capacity(dependencies.len());
    let dep_graph = dependencies_to_graph(dependencies);
    let mut dfs = Dfs::new(&dep_graph, node_index(0));

    while let Some(node) = dfs.next(&dep_graph) {
        projective_order.push(node);
    }

    // Now create an array that specifies for each token its 'rank'
    // in the projective order. Fixme: there is the assumption here
    // that every token occurs in the dependency graph.
    let mut order_mapping = vec![0; projective_order.len()];
    for (idx, token) in projective_order.iter().enumerate() {
        order_mapping[token.index()] = idx
    }

    order_mapping
}
